pkg/keyspace/tso_keyspace_group.go

```diff
@@ -16,47 +16,97 @@ package keyspace
 	"github.com/tikv/pd/pkg/mcs/utils"
+	"github.com/tikv/pd/pkg/slice"
 	"github.com/tikv/pd/pkg/storage/endpoint"
 	"github.com/tikv/pd/pkg/storage/kv"
 )
 
+const (
+	opAdd    = "add"
+	opDelete = "delete"
+)
+
 // GroupManager is the manager of keyspace group related data.
 type GroupManager struct {
 	ctx context.Context
+	sync.RWMutex
+	// groups is the cache of keyspace group related information.
+	// user kind -> keyspace group
+	groups map[endpoint.UserKind]*indexedHeap
 	// store is the storage for keyspace group related information.
 	store endpoint.KeyspaceGroupStorage
 }
 
 // NewKeyspaceGroupManager creates a Manager of keyspace group related data.
 func NewKeyspaceGroupManager(ctx context.Context, store endpoint.KeyspaceGroupStorage) *GroupManager {
+	groups := make(map[endpoint.UserKind]*indexedHeap)
+	for i := 0; i < int(endpoint.UserKindCount); i++ {
+		groups[endpoint.UserKind(i)] = newIndexedHeap(int(utils.MaxKeyspaceGroupCountInUse))
+	}
 	return &GroupManager{
-		ctx:   ctx,
-		store: store,
+		ctx:    ctx,
+		store:  store,
+		groups: groups,
 	}
 }
 
 // Bootstrap saves default keyspace group info.
 func (m *GroupManager) Bootstrap() error {
 	defaultKeyspaceGroup := &endpoint.KeyspaceGroup{
-		ID: utils.DefaultKeySpaceGroupID,
-		// TODO: define a user kind type
-		UserKind: "default",
+		ID:       utils.DefaultKeySpaceGroupID,
+		UserKind: endpoint.Basic.String(),
 	}
-	err := m.saveKeyspaceGroups([]*endpoint.KeyspaceGroup{defaultKeyspaceGroup})
-	// It's possible that default keyspace group already exists in the storage (e.g. PD restart/recover),
-	// so we ignore the ErrKeyspaceGroupExists.
+
+	// Ignore the error if default keyspace group already exists in the storage (e.g. PD restart/recover).
+	err := m.saveKeyspaceGroups([]*endpoint.KeyspaceGroup{defaultKeyspaceGroup}, false)
 	if err != nil && err != ErrKeyspaceGroupExists {
 		return err
 	}
 
+	m.Lock()
+	defer m.Unlock()
+	userKind := endpoint.StringUserKind(defaultKeyspaceGroup.UserKind)
+
+	// If the group for the userKind does not exist, create a new one.
+	if _, ok := m.groups[userKind]; !ok {
+		m.groups[userKind] = newIndexedHeap(int(utils.MaxKeyspaceGroupCountInUse))
+	}
+	m.groups[userKind].Put(defaultKeyspaceGroup)
+
+	// Load all the keyspace groups from the storage and add to the respective userKind groups.
+	groups, err := m.store.LoadKeyspaceGroups(utils.DefaultKeySpaceGroupID, 0)
+	if err != nil {
+		return err
+	}
+	for _, group := range groups {
+		userKind := endpoint.StringUserKind(group.UserKind)
+		if _, ok := m.groups[userKind]; !ok {
+			m.groups[userKind] = newIndexedHeap(int(utils.MaxKeyspaceGroupCountInUse))
+		}
+		m.groups[userKind].Put(group)
+	}
+
 	return nil
 }
 
 // CreateKeyspaceGroups creates keyspace groups.
 func (m *GroupManager) CreateKeyspaceGroups(keyspaceGroups []*endpoint.KeyspaceGroup) error {
-	return m.saveKeyspaceGroups(keyspaceGroups)
+	if err := m.saveKeyspaceGroups(keyspaceGroups, false); err != nil {
```

Comment:
Do we need read lock to protect keyspaceGroups from being concurrently accessed in m.saveKeyspaceGroups?

Reply:
I added a lock in group manager level, which can prevent saveKeyspaceGroups from calling concurrently.